docs/Client.md

```diff
@@ -0,0 +1,1011 @@
+
+#### Example 1 - Basic GET stream request
+
+```js
+'use strict'
+const { createServer } = require('http')
+const { Client } = require('undici')
+const { Writable } = require('stream')
+
+const server = createServer((request, response) => {
+  response.end('Hello, World!')
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  const bufs = []
+
+  client.stream({
+    path: '/',
+    method: 'GET',
+    opaque: { bufs }
+  }, ({ statusCode, headers, opaque: { bufs } }) => {
+    console.log(`response received ${statusCode}`)
+    console.log('headers', headers)
+    return new Writable({
+      write (chunk, encoding, callback) {
+        bufs.push(chunk)
+        callback()
+      }
+    })
+  }).then(({ opaque: { bufs } }) => {
+    console.log(Buffer.concat(bufs).toString('utf-8'))
+
+    client.close()
+    server.close()
+  }).catch(error => {
+    console.error(error)
+  })
+})
+```
+
+#### Example 2 - Stream to Fastify Response
+
+In this example, a (fake) request is made to the fastify server using `fastify.inject()`. This request then executes the fastify route handler which makes a subsequent request to the raw Node.js http server using `undici.client.stream()`. The fastify response is passed to the `opaque` option so that undici can tap into the underlying writable stream using `response.raw`. This methodology demonstrates how one could use undici and fastify together to create fast-as-possible requests from one backend server to another.
+
+```js
+'use strict'
+
+const { createServer } = require('http')
+const undici = require('undici')
+const fastify = require('fastify')
+
+const nodeServer = createServer((request, response) => {
+  response.end('Hello, World! From Node.js HTTP Server')
+})
+
+nodeServer.listen(() => {
+  console.log('Node Server listening')
+
+  const nodeServerUndiciClient = new undici.Client(`http://localhost:${nodeServer.address().port}`)
+
+  const fastifyServer = fastify()
+
+  fastifyServer.route({
+    url: '/',
+    method: 'GET',
+    handler: (request, response) => {
+      nodeServerUndiciClient.stream({
+        path: '/',
+        method: 'GET',
+        opaque: response
+      }, ({ opaque }) => opaque.raw)
+    }
+  })
+
+  fastifyServer
+    .listen()
+    .then(() => {
+      console.log('Fastify Server listening')
+      const fastifyServerUndiciClient = new undici.Client(`http://localhost:${fastifyServer.server.address().port}`)
+
+      fastifyServerUndiciClient.request({
+        path: '/',
+        method: 'GET'
+      }).then(({ statusCode, body }) => {
+        console.log(`response received ${statusCode}`)
+        body.setEncoding('utf8')
+        body.on('data', console.log)
+
+        nodeServerUndiciClient.close()
+        fastifyServerUndiciClient.close()
+        fastifyServer.close()
+        nodeServer.close()
+      })
+    })
+})
+```
+
+### `Client.upgrade(options[, callback])`
+
+Upgrade the client to a different protocol.
```

Comment:
Reference to https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism